SU2_CFD/src/solver_direct_elasticity.cpp

```diff
@@ -2160,27 +2166,16 @@ void CFEASolver::Compute_NodalStress(CGeometry *geometry, CSolver **solver_conta
         case (NEWMARK_IMPLICIT):
           
           /*--- Loop over all points, and set aux vector TimeRes_Aux = a0*U+a2*U'+a3*U'' ---*/
-          if (linear_analysis) {
```

Comment:
Hi Ruben,
Thanks for pointing that out, I was able to reproduce the issue on my side. Looks like I opened I nice can of worms for myself :) but reverting that section makes the problem worse.
On the subject of cans of worms, just after that section there's this:
/*--- Once computed, compute M*TimeRes_Aux ---*/
MassMatrix.MatrixVectorProduct(TimeRes_Aux,TimeRes,geometry,config);
Which I think will break differentiation w.r.t. density in dynamic cases as CSysMatrix::AddBlock uses SU2_TYPE::GetValue on the input data.
Cheers,
Pedro

Reply:
Hi Pedro, 
I guess you will need to go back to the definition of the Neumann scheme and start from there with your residual approach to make sure the time-domain computation is done correctly. From that definition it's where the difference in treatment for linear and non-linear came.

About your other comment, yes, that's 100% right. The `MatrixVectorProduct` loses track of the dependencies for time-domain problems. That's why if you go to `CSysMatrix::SetBlock`, `CSysMatrix::AddBlock` and `CSysMatrix::SubtractBlock` you'll find a commented version of the routines in which the matrix structure stores the full val_block rather than the value alone. 

I implemented it to test dynamic structural adjoints (which work for time-averaged objective functions if you uncomment those lines), but left it out when merging with develop as this increases the memory footprint elsewhere in the code, awaiting to see if a better solution could be found.

Cheers,
Ruben

Reply:
Understood.
One option would be to compute the "time residual" like I am computing the residual now, that would be expensive as it would involve recomputing the mass matrix every time.
Another option would be to template the matrix class with / without AD support. The without version could even use passivedouble thereby taking advantage of any BLAS optimizations even in AD mode.
Anyway, this is me spitballing while probably overlooking some other aspects...